File: wordle_solver.py
import os
from wordfreq import word_frequency

def load_words(filename):
    script_dir = os.path.dirname(os.path.abspath(__file__))
    file_path = os.path.join(script_dir, filename)
    with open(file_path, 'r') as file:
        words = [line.strip() for line in file]
    return words

valid_words = load_words("valid_words.txt")

def filter_words(words, filters):
    filtered_words = words
    for func, args in filters:
        filtered_words = func(filtered_words, *args)
    return list(filtered_words)

def filter_words_by_min_count(words, letter, min_count):
    return (word for word in words if word.count(letter) >= min_count)

def filter_words_by_exact_count(words, letter, exact_count):
    return (word for word in words if word.count(letter) == exact_count)

def filter_words_by_position(words, letter, position):
    return (word for word in words if word[position] == letter)

def filter_words_by_exclusion_at(words, letter, position):
    return (word for word in words if word[position] != letter)

def filter_words_by_exclusion(words, letter):
    return (word for word in words if letter not in word)

def count_letter_frequencies(words):
    frequencies = {}
    for word in words:
        word_freq = {}
        for char in word:
            if char in word_freq:
                word_freq[char] += 1
            else:
                word_freq[char] = 1
        for char, count in word_freq.items():
            if char in frequencies:
                frequencies[char] += count
            else:
                frequencies[char] = count
    return frequencies

def get_letter_counts_from_guesses(guesses):
    letter_counts = {}
    for word, feedback in guesses:
        temp_counts = {}
        for index, char in enumerate(word):
            if feedback[index] in {'g', 'y'}:
                if char in temp_counts:
                    temp_counts[char] += 1
                else:
                    temp_counts[char] = 1
        for char in temp_counts:
            if char in letter_counts:
                letter_counts[char] = max(letter_counts[char], temp_counts[char])
            else:
                letter_counts[char] = temp_counts[char]
    return letter_counts

def get_green_letters_from_guesses(guesses):
    green_letters = {}
    for word, feedback in guesses:
        for index, char in enumerate(word):
            if feedback[index] == 'g':
                green_letters[index] = char
    return green_letters

def get_restrictions_from_guesses(guesses):
    excluded_letters = set()
    exact_counts = {}
    position_restrictions = {}
    for word, feedback in guesses:
        occurrences = {}
        for index, char in enumerate(word):
            if char not in occurrences:
                occurrences[char] = {}
            occurrences[char][index] = feedback[index]
        for char, feedbacks in occurrences.items():
            feedback_types = set(feedbacks.values())
            if 'b' in feedback_types:
                count = sum(1 for feedback in feedbacks.values() if feedback in {'g', 'y'})
                exact_counts[char] = count
                excluded_positions = {i for i, feedback in feedbacks.items() if feedback in {'b', 'y'}}
                if char not in position_restrictions:
                    position_restrictions[char] = set()
                position_restrictions[char].update(excluded_positions)
            else:
                excluded_positions = {i for i, feedback in feedbacks.items() if feedback == 'y'}
                if char not in position_restrictions:
                    position_restrictions[char] = set()
                position_restrictions[char].update(excluded_positions)
    return excluded_letters, exact_counts, position_restrictions

def calculate_frequencies(words):
    frequencies = {}
    total_frequency = 0
    for word in words:
        freq = word_frequency(word, 'en')
        frequencies[word] = freq
        total_frequency += freq
    return {word: freq / total_frequency for word, freq in frequencies.items()}

def expected_remaining_words(words, frequencies):
    expected_counts = {}
    for guess in words:
        total_count = 0
        for answer in words:
            probability = frequencies.get(answer, 0)
            remaining_count = count_remaining_words_after_guess(guess, answer, words)
            total_count += remaining_count * probability
        expected_counts[guess] = total_count
    return expected_counts

def count_remaining_words_after_guess(guess, answer, words):
    feedback = generate_feedback(guess, answer)
    filtered_words = filter_words(words, [(filter_words_by_feedback, (guess, feedback))])
    return len(filtered_words)

def generate_feedback(guess, answer):
    feedback = {}
    for index in range(5):
        if guess[index] == answer[index]:
            feedback[index] = "g"
        elif guess[index] not in answer:
            feedback[index] = "b"
    if len(feedback) == 5:
        return ''.join(feedback.get(i, "b") for i in range(5))
    filtered_guess = {i: char for i, char in enumerate(guess) if i not in feedback}
    filtered_answer = [char for i, char in enumerate(answer) if i not in feedback or feedback.get(i) != "g"]
    for index in range(5):
        if index not in feedback:
            char = filtered_guess.get(index)
            if char in filtered_answer:
                feedback[index] = "y"
                filtered_answer.remove(char)
            else:
                feedback[index] = "b"
    return ''.join(feedback.get(i, "b") for i in range(5))

def filter_words_by_feedback(words, guess, feedback):
    remaining_words = filter_words(words, [
        (filter_words_by_min_count, (char, count))
        for char, count in get_letter_counts_from_guesses([(guess, feedback)]).items()
    ])
    remaining_words = filter_words(remaining_words, [
        (filter_words_by_position, (letter, index))
        for index, letter in get_green_letters_from_guesses([(guess, feedback)]).items()
    ])
    remaining_words = filter_words(remaining_words, [
        (filter_words_by_exclusion, (letter,))
        for letter in get_restrictions_from_guesses([(guess, feedback)])[0]
    ])
    remaining_words = filter_words(remaining_words, [
        (filter_words_by_exact_count, (letter, count))
        for letter, count in get_restrictions_from_guesses([(guess, feedback)])[1].items()
    ])
    remaining_words = filter_words(remaining_words, [
        (filter_words_by_exclusion_at, (letter, position))
        for letter, positions in get_restrictions_from_guesses([(guess, feedback)])[2].items()
        for position in positions
    ])
    return remaining_words

def find_best_guesses(remaining_words, top_n=3):
    frequencies = calculate_frequencies(remaining_words)
    expected_words = expected_remaining_words(remaining_words, frequencies)
    sorted_words = sorted(expected_words, key=expected_words.get)
    return sorted_words[:top_n]

File: src/wordle_solver.cpp
#include "wordle_solver.hpp"
#include <algorithm>
#include <cmath>
#include <unordered_map>
#include <unordered_set>
#include <numeric>

// Helper function implementations
std::unordered_map<char, int> getLetterCountsFromGuesses(const std::vector<std::pair<std::string, std::string>>& guesses) {
    std::unordered_map<char, int> counts;
    for (const auto& [guess, feedback] : guesses) {
        for (size_t i = 0; i < guess.length(); ++i) {
            if (feedback[i] != 'b') {
                counts[guess[i]]++;
            }
        }
    }
    return counts;
}

std::unordered_map<int, char> getGreenLettersFromGuesses(const std::vector<std::pair<std::string, std::string>>& guesses) {
    std::unordered_map<int, char> greenLetters;
    for (const auto& [guess, feedback] : guesses) {
        for (size_t i = 0; i < guess.length(); ++i) {
            if (feedback[i] == 'g') {
                greenLetters[i] = guess[i];
            }
        }
    }
    return greenLetters;
}

std::tuple<std::vector<char>, std::unordered_map<char, int>, std::vector<std::pair<int, char>>> 
getRestrictionsFromGuesses(const std::vector<std::pair<std::string, std::string>>& guesses) {
    std::vector<char> excludedLetters;
    std::unordered_map<char, int> exactCounts;
    std::vector<std::pair<int, char>> positionRestrictions;

    for (const auto& [guess, feedback] : guesses) {
        for (size_t i = 0; i < guess.length(); ++i) {
            if (feedback[i] == 'b') {
                excludedLetters.push_back(guess[i]);
            } else if (feedback[i] == 'y') {
                positionRestrictions.emplace_back(i, guess[i]);
            }
        }
    }

    return {excludedLetters, exactCounts, positionRestrictions};
}

std::vector<std::string> filterWordsByExclusion(const std::vector<std::string>& words, const std::vector<char>& excludedLetters) {
    std::vector<std::string> filtered;
    std::copy_if(words.begin(), words.end(), std::back_inserter(filtered),
        [&](const std::string& word) {
            for (char c : excludedLetters) {
                if (word.find(c) != std::string::npos) return false;
            }
            return true;
        });
    return filtered;
}

std::vector<std::string> filterWordsByPosition(const std::vector<std::string>& words, const std::unordered_map<int, char>& greenLetters) {
    std::vector<std::string> filtered;
    std::copy_if(words.begin(), words.end(), std::back_inserter(filtered),
        [&](const std::string& word) {
            for (const auto& [pos, c] : greenLetters) {
                if (word[pos] != c) return false;
            }
            return true;
        });
    return filtered;
}

std::vector<std::string> filterWordsByExactCount(const std::vector<std::string>& words, const std::unordered_map<char, int>& exactCounts) {
    std::vector<std::string> filtered;
    std::copy_if(words.begin(), words.end(), std::back_inserter(filtered),
        [&](const std::string& word) {
            std::unordered_map<char, int> counts;
            for (char c : word) counts[c]++;
            for (const auto& [c, count] : exactCounts) {
                if (counts[c] != count) return false;
            }
            return true;
        });
    return filtered;
}

std::vector<std::string> filterWordsByExclusionAt(const std::vector<std::string>& words, const std::vector<std::pair<int, char>>& positionRestrictions) {
    std::vector<std::string> filtered;
    std::copy_if(words.begin(), words.end(), std::back_inserter(filtered),
        [&](const std::string& word) {
            for (const auto& [pos, c] : positionRestrictions) {
                if (word[pos] == c) return false;
            }
            return true;
        });
    return filtered;
}

// Core functions from your original code
std::vector<std::string> filterWordsByFeedback(const std::vector<std::string>& words, const std::string& guess, const std::string& feedback) {
    auto letterCounts = getLetterCountsFromGuesses({{guess, feedback}});
    auto greenLetters = getGreenLettersFromGuesses({{guess, feedback}});
    auto [excludedLetters, exactCounts, positionRestrictions] = getRestrictionsFromGuesses({{guess, feedback}});

    std::vector<std::string> filteredWords = words;
    filteredWords = filterWordsByExclusion(filteredWords, excludedLetters);
    filteredWords = filterWordsByPosition(filteredWords, greenLetters);
    filteredWords = filterWordsByExactCount(filteredWords, exactCounts);
    filteredWords = filterWordsByExclusionAt(filteredWords, positionRestrictions);

    return filteredWords;
}

std::vector<std::string> findBestGuesses(const std::vector<std::string>& remainingWords, int topN) {
    auto frequencies = calculateFrequencies(remainingWords);
    auto expectedWords = expectedRemainingWords(remainingWords, frequencies);
    std::vector<std::string> sortedWords(remainingWords);
    
    std::sort(sortedWords.begin(), sortedWords.end(), 
        [&expectedWords](const std::string& a, const std::string& b) {
            return expectedWords.at(a) < expectedWords.at(b);
        });
        
    if (sortedWords.size() > static_cast<size_t>(topN)) {
        sortedWords.resize(topN);
    }
    return sortedWords;
}

std::unordered_map<std::string, double> calculateFrequencies(const std::vector<std::string>& words) {
    std::unordered_map<std::string, double> frequencies;
    double totalFrequency = 0.0;
    for (const auto& word : words) {
        double freq = 1.0;
        frequencies[word] = freq;
        totalFrequency += freq;
    }
    for (auto& [word, freq] : frequencies) {
        freq /= totalFrequency;
    }
    return frequencies;
}

std::unordered_map<std::string, double> expectedRemainingWords(
    const std::vector<std::string>& words,
    const std::unordered_map<std::string, double>& frequencies) {
    
    std::unordered_map<std::string, double> expectedCounts;
    for (const auto& guess : words) {
        double totalCount = 0.0;
        for (const auto& answer : words) {
            double probability = frequencies.at(answer);
            int remainingCount = countRemainingWordsAfterGuess(guess, answer, words);
            totalCount += remainingCount * probability;
        }
        expectedCounts[guess] = totalCount;
    }
    return expectedCounts;
}

int countRemainingWordsAfterGuess(const std::string& guess, const std::string& answer, const std::vector<std::string>& words) {
    std::string feedback = generateFeedback(guess, answer);
    std::vector<std::string> filteredWords = filterWordsByFeedback(words, guess, feedback);
    return filteredWords.size();
}

std::string generateFeedback(const std::string& guess, const std::string& answer) {
    std::string feedback(guess.size(), 'b');
    
    // First pass: Mark green letters
    for (size_t i = 0; i < guess.size(); ++i) {
        if (guess[i] == answer[i]) {
            feedback[i] = 'g';
        }
    }
    
    // Second pass: Mark yellow letters
    std::unordered_map<char, int> remainingLetters;
    for (size_t i = 0; i < answer.size(); ++i) {
        if (feedback[i] != 'g') {
            remainingLetters[answer[i]]++;
        }
    }
    
    for (size_t i = 0; i < guess.size(); ++i) {
        if (feedback[i] == 'b' && remainingLetters[guess[i]] > 0) {
            feedback[i] = 'y';
            remainingLetters[guess[i]]--;
        }
    }
    
    return feedback;
}
